File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Controllers/BookstoreProductController.cs
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using FluentAssertions;
using RekindleTaf.Configuration;
using RekindleTaf.Configuration.Rest;
using RekindleTaf.Model;
using RestSharp;

namespace RekindleTaf.Controllers;

public class BookstoreProductController(IEnvConfig envConfig, IRestConfig restConfig) : IBookstoreProductController
{
    public async Task<RestResponse> Get(string param = "")
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresProductById)
            .AddUrlSegment("productId", param);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public Product GetSuccessfully(string param = "")
    {
        var response = Get(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<Product>(response);
    }

    public async Task<RestResponse> GetAll(string param = "")
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresProducts);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public List<Product> GetAllSuccessfully(string param = "")
    {
        var response = GetAll(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<List<Product>>(response);
    }

    public async Task<RestResponse> Post(Product param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoreProductByStoreId, Method.Post)
            .AddUrlSegment("bookstoreId", param.BookstoreId)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePostAsync(request);
    }

    public Product PostSuccessfully(Product param)
    {
        var response = Post(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.Created);
        var responseData = restConfig.Deserialize<JsonNode>(response);
        param.Id = responseData.ToString();
        return param;
    }

    public async Task<RestResponse> Put(Product param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresProductById, Method.Put)
            .AddUrlSegment("productId", param.Id)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePutAsync(request);
    }

    public void PutSuccessfully(Product param)
    {
        var response = Put(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NoContent);
    }

    public async Task<RestResponse> Delete(Product param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoreProductByStoreIdAndProductId,
                Method.Delete)
            .AddUrlSegment("bookstoreId", param.BookstoreId)
            .AddUrlSegment("productId", param.Id);
        return await restConfig.RekindleClient().ExecuteDeleteAsync(request);
    }

    public void DeleteSuccessfully(Product param)
    {
        var response = Delete(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NoContent);
    }
}

File: typescript-project/rekindle-book-store-aqa/src/rest/controllers/bookstore-product-controller.ts
import {Product} from "@/model/domain-model";
import {AxiosInstance, AxiosResponse, HttpStatusCode} from "axios";
import {BookstoreEndpoints} from "@/rest/Endpoint";
import {BookstoreProductCrudController} from "@/rest/controllers/rest-controllers";
import {expect} from "@jest/globals";
import {RekindleClient} from "@/rest/rekindle-rest-client";

export class BookstoreProductController implements BookstoreProductCrudController {

    private readonly client: AxiosInstance;

    constructor(client: RekindleClient) {
        this.client = client.getAxios();
    }

    async get(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: BookstoreEndpoints.bookstoresProductById({productId: param})
        });
    }

    async getSuccessfully(param?: string): Promise<Product> {
        try {
            const resp = await this.get(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data;
        } catch (err) {
            throw err;
        }
    }

    async getAll(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: BookstoreEndpoints.bookstoresProducts
        });
    }

    async getAllSuccessfully(param?: string): Promise<Product[]> {
        try {
            const resp = await this.getAll(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data
        } catch (err) {
            throw err;
        }
    }

    async post(param: Product): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'POST',
            url: BookstoreEndpoints.bookstoreProductByStoreId({bookstoreId: param.bookstoreId}),
            data: param
        });
    }

    async postSuccessfully(param: Product): Promise<Product> {
        try {
            let bookstore = param;
            const resp = await this.post(param);
            expect(resp.status).toBe(HttpStatusCode.Created);
            bookstore.id = resp.data
            return bookstore;
        } catch (err) {
            throw err;
        }
    }

    async put(param: Product): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'PUT',
            url: BookstoreEndpoints.bookstoresProductById({productId: param.id}),
            data: param
        });
    }

    async putSuccessfully(param: Product): Promise<void> {
        try {
            const resp = await this.put(param);
            expect(resp.status).toBe(HttpStatusCode.NoContent);
        } catch (err) {
            throw err;
        }
    }

    async delete(param: Product): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'DELETE',
            url: BookstoreEndpoints.bookstoreProductByStoreIdAndProductId({
                bookstoreId: param.bookstoreId,
                productId: param.id
            })
        });
    }

    async deleteSuccessfully(param: Product): Promise<void> {
        try {
            const resp = await this.delete(param);
            expect(resp.status).toBe(HttpStatusCode.NoContent);
        } catch (err) {
            throw err;
        }
    }
}